src/wayland/data_device/server_dnd_grab.rs

```diff
@@ -182,147 +189,170 @@ where
-                    return;
-                }
-                (&mut *callback.borrow_mut())(ServerDndEvent::Finished);
-                data.active = false;
+        }
+        Request::Destroy => {}
+        Request::Finish => {
+            if !data.active {
+                offer.post_error(
+                    dh,
+                    wl_data_offer::Error::InvalidFinish as u32,
+                    "Cannot finish a data offer that is no longer active.",
+                );
+                return;
             }
-            Request::SetActions {
-                dnd_actions,
-                preferred_action,
-            } => {
-                let preferred_action = preferred_action;
-
-                // preferred_action must only contain one bitflag at the same time
-                if ![DndAction::None, DndAction::Move, DndAction::Copy, DndAction::Ask]
-                    .contains(&preferred_action)
-                {
-                    offer.as_ref().post_error(
-                        wl_data_offer::Error::InvalidAction as u32,
-                        "Invalid preferred action.".into(),
-                    );
-                    return;
-                }
-                let possible_actions = metadata.dnd_action & dnd_actions;
-                data.chosen_action = (&mut *action_choice.borrow_mut())(possible_actions, preferred_action);
-                // check that the user provided callback respects that one precise action should be chosen
-                debug_assert!(
-                    [DndAction::None, DndAction::Move, DndAction::Copy, DndAction::Ask]
-                        .contains(&data.chosen_action)
+            if !data.accepted {
+                offer.post_error(
+                    dh,
+                    wl_data_offer::Error::InvalidFinish,
+                    "Cannot finish a data offer that has not been accepted.",
+                );
+                return;
+            }
+            if !data.dropped {
+                offer.post_error(
+                    dh,
+                    wl_data_offer::Error::InvalidFinish,
+                    "Cannot finish a data offer that has not been dropped.",
+                );
+                return;
+            }
+            if data.chosen_action.is_empty() {
+                offer.post_error(
+                    dh,
+                    wl_data_offer::Error::InvalidFinish,
+                    "Cannot finish a data offer with no valid action.",
                 );
-                offer.action(data.chosen_action);
-                (&mut *callback.borrow_mut())(ServerDndEvent::Action(data.chosen_action));
+                return;
             }
-            _ => unreachable!(),
+            (*callback.lock().unwrap())(ServerDndEvent::Finished);
+            data.active = false;
         }
-    });
+        Request::SetActions {
+            dnd_actions,
+            preferred_action,
+        } => {
+            let dnd_actions = match dnd_actions {
```

Comment:
Could be `into_result` and then `unwrap_or`

Reply:
Oh, that method used to not exist before, that's nice!